digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/ProjectValidator.scala

```diff
@@ -576,28 +576,35 @@ class ProjectValidator {
       } else None
     }
 
-    def checkMinMaxTrackRoadTypes(trackInterval: Seq[ProjectLink]): Option[ProjectLink] = {
+    def checkMinMaxTrackRoadTypes(trackInterval: Seq[ProjectLink]): Option[Seq[ProjectLink]] = {
       val (left, right) = trackInterval.partition(_.track == Track.LeftSide)
 
-      val leftRoadTypes = left.sortBy(_.startAddrMValue).map(_.roadType.value).foldLeft(Seq.empty[Int]) { case (list, next) =>
-        if (list.nonEmpty && list.last == next)
-          list
+      val leftRoadTypes = left.sortBy(_.startAddrMValue).foldLeft(Seq.empty[(Int, Long, Long)]) { case (list, next) =>
+        if (list.nonEmpty && list.last._1 == next.roadType.value)
+          list.init :+ list.last.copy(_3 = next.endAddrMValue)
         else
-          list :+ next
+          list :+ (next.roadType.value, next.startAddrMValue, next.endAddrMValue)
       }
 
-      val rightRoadTypes = right.sortBy(_.startAddrMValue).map(_.roadType.value).foldLeft(Seq.empty[Int]) { case (list, next) =>
-        if (list.nonEmpty && list.last == next)
-          list
+      val rightRoadTypes = right.sortBy(_.startAddrMValue).foldLeft(Seq.empty[(Int, Long, Long)]) { case (list, next) =>
+        if (list.nonEmpty && list.last._1 == next.roadType.value)
+          list.init :+ list.last.copy(_3 = next.endAddrMValue)
         else
-          list :+ next
+          list :+ (next.roadType.value, next.startAddrMValue, next.endAddrMValue)
       }
 
-      if (!(leftRoadTypes sameElements rightRoadTypes)) {
-        if (left.nonEmpty)
-          Some(left.head)
-        else
-          Some(right.head)
+      if (!(leftRoadTypes equals rightRoadTypes)) {
```

Comment:
Now we are checking links two times in case there are different amount of links on different tracks. Would it be enough to use this second algorithm to check all cases: roads with equal amount of links on both tracks and roads with different amount of links on different tracks?

Reply:
Done.